falcon/util/misc.py

```diff
@@ -373,7 +376,16 @@ def secure_filename(filename: str) -> str:
     filename = unicodedata.normalize('NFKD', filename)
     if filename.startswith('.'):
         filename = filename.replace('.', '_', 1)
-    return _UNSAFE_CHARS.sub('_', filename)
+    filename = _UNSAFE_CHARS.sub('_', filename)
+
+    if max_length is not None and len(filename) > max_length:
+        name, ext = os.path.splitext(filename)
+        if max_length - len(ext) >= 1:
+            filename = name[: max_length - len(ext)] + ext
+        else:
+            filename = filename[:max_length]
+
+    return filename
```

Comment:
as a side note I noticed that we don't hold windows hand regarding device file names, like
https://github.com/pallets/werkzeug/blob/7868bef5d978093a8baa0784464ebe5d775ae92a/src/werkzeug/utils.py#L229-L237

this is likely another issue / pr though

Reply:
Yes, let's create a separate issue if this is desirable. Not really sure this belongs to the framework either. Normally the filename should be appended to a path, is it still an issue on Windows if you write like `C:\Uploads\NUL`?

Reply:
Yes that's not valid in any folder

Reply:
I created an issue from your comment: #2422.